### PR TITLE
[cglm] add cglm

### PR DESCRIPTION
**Pull request by suyashmohan**
**Add package for cglm**

- What does your PR fix? 
Add support for cglm which is a Highly Optimized Graphics Math (glm) lib for C

### CODE REVIEW

File: ports/cglm/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,29 @@
+include(vcpkg_common_functions)
+
```
Comment (PhoebeHui) [inline ports/cglm/portfile.cmake:2]: Could you please remove this? it's deprecated.

**Review by PhoebeHui — CHANGES_REQUESTED** [review]
The cglm failed on CI testing, could you have a look? 

Starting package 1/1: cglm:x64-windows
Building package cglm[core]:x64-windows...
Could not locate cached archive: W:\59\596dc1de4f00e28d8c2f4bc265739a340e84b53c.zip
-- Downloading https://github.com/recp/cglm/archive/v0.7.8.tar.gz...
-- Extracting source D:/downloads/recp-cglm-v0.7.8.tar.gz
-- Using source at D:/buildtrees/cglm/src/v0.7.8-ef8c98816b.clean
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: D:/packages/cglm_x64-windows/share/cglm/copyright
-- Performing post-build validation

The following dlls were found in /lib or /debug/lib. Please move them to /bin or /debug/bin, respectively.

    D:/packages/cglm_x64-windows/lib/cglm.dll


The following dlls were found in /lib or /debug/lib. Please move them to /bin or /debug/bin, respectively.

    D:/packages/cglm_x64-windows/debug/lib/cglm.dll

Found 2 error(s). Please correct the portfile:
    C:\a\1\s\ports\cglm\portfile.cmake
-- Performing post-build validation done
Error: Building package cglm:x64-windows failed with: POST_BUILD_CHECKS_FAILED
Elapsed time for package cglm:x64-windows: 11.33 s

**Comment by PhoebeHui** [comment]
The cglm:arm_uwp failed with error C2719 on CI testing, could you please have a look?

```
D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(47,39): error C2719: 'd': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
     4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(63,34): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
     4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(79,32): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
     4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(94,36): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
     4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(108,25): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
     4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(120,30): error C2719: 'a': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
...
```

**Comment by suyashmohan** [comment]
> The cglm:arm_uwp failed with error C2719 on CI testing, could you please have a look?
> 
> ```
> D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(47,39): error C2719: 'd': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
>      4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(63,34): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
>      4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(79,32): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
>      4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(94,36): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
>      4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(108,25): error C2719: 'v': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
>      4>D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\include\cglm\vec4-ext.h(120,30): error C2719: 'a': formal parameter with requested alignment of 16 won't be aligned (compiling source file D:\buildtrees\cglm\src\v0.7.8-ef8c98816b.clean\src\euler.c) [D:\buildtrees\cglm\arm-uwp-dbg\cglm.vcxproj]
> ...
> ```

I am not sure if cglm is support on arm for uwp or not. So I have added following in CONTROL file to ignore arm and uwp

```
Supports: !(uwp|arm)
```

Is that enough?

File: ports/cglm/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/cglm RENAME copyright)
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib/cglm")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/lib/cglm")
+configure_file(${CMAKE_CURRENT_LIST_DIR}/cglm-config.cmake  ${CURRENT_PACKAGES_DIR}/share/cglm COPYONLY)
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
+endif()
```
Comment (PhoebeHui) [inline ports/cglm/portfile.cmake:31]:
```suggestion
```

**Review by PhoebeHui — CHANGES_REQUESTED** [review]
@suyashmohan, it's fine to add them to 'Supports'.

Could your address the following review suggestions? except this, others looks good to me.

**Comment by PhoebeHui** [comment]
ping @suyashmohan..

**Comment by suyashmohan** [comment]
> ping @suyashmohan..

Sorry, missed it. Will update the PR this week :)

**Comment by suyashmohan** [comment]
@PhoebeHui I have applied one of your suggestions. But suggestions regarding declaring  CGLM_SHARED and CGLM_STATIC, I want to force building static build all the time, hence haven't applied them. The original cmake file from cglm builds dll in  /lib or /debug/lib directories. This is not allowed in vcpkg, DLLs should be in bin and debug/bin directories.  I tried to move the DLLs in bin directory, but when we use cglm library in a project, the original cmake from cglm fails as it checks for DLLs in lib folder. If we need shared dll build support in this vcpkg, then we need to modify the cmake file in cglm repo, which is out of scope from this PR.

**Comment by PhoebeHui** [comment]
@suyashmohan, you may need to add 'vcpkg_check_linkage(ONLY_STATIC_LIBRARY)' to portfile.cmake.

@dan-shaw, could you help further review?

**Comment by suyashmohan** [comment]
@dan-shaw @PhoebeHui This PR seems to be stuck for 20 days, any conclusions that we can make? 😅

**Comment by PhoebeHui** [comment]
@suyashmohan, sorry for my late response. 

We will take this in tommorw meeting, and get back to you as soon as possible.

**Comment by BillyONeal** [comment]
> @PhoebeHui I have applied one of your suggestions. But suggestions regarding declaring CGLM_SHARED and CGLM_STATIC, I want to force building static build all the time, hence haven't applied them.

Whether libraries build static generally that's a decision the triplet makes where possible. The presence of both these variables suggests that the project supports both and we should respect that. If you personally want a static build then you should use a static triplet.

**Comment by suyashmohan** [comment]
> > @PhoebeHui I have applied one of your suggestions. But suggestions regarding declaring CGLM_SHARED and CGLM_STATIC, I want to force building static build all the time, hence haven't applied them.
> 
> Whether libraries build static generally that's a decision the triplet makes where possible. The presence of both these variables suggests that the project supports both and we should respect that. If you personally want a static build then you should use a static triplet.

@BillyONeal Thanks for responding. The issue is this library's own cmake checks for DLL file presence in certain path like 'lib' or 'debug/lib'. vcpkg doesn't allow this path. So either there has to be change in cmake in original cglm repo or we find a way to add exception to vcpkg's behaviour for this package. I have no control over original cglm cmake.

**Comment by BillyONeal** [comment]
>So either there has to be change in cmake in original cglm repo or we find a way to add exception to vcpkg's behaviour for this package. I have no control over original cglm cmake.

You can author a patch:

1. `vcpkg install cglm --editable`
2. `cd ` (wherever the source got extracted) (C:\Dev\vcpkg2\buildtrees\cglm\src\v0.7.8-ef8c98816b for me)
3. `git init`
4. `git add -A`
5. `git commit -m initial`
6. (make whatever changes you want)
7. Open a new window back in the vcpkg root and `vcpkg remove cglm`
8. (also in the vcpkg root) `vcpkg install cglm --editable` (and check that that works)
9. (back in the cglm source tree) `git diff > path/to/vcpkg/ports/cglm/fix-directories.patch`
10. Add that patch to `PATCHES` in `vcpkg_from_github`
11. (back in the vcpkg prompt again) check that it worked with `vcpkg remove cglm && vcpkg install cglm` (without `--editable` this time)

Example: https://github.com/suyashmohan/vcpkg/pull/1

**Comment by BillyONeal** [comment]
Here are shinier instructions: https://github.com/microsoft/vcpkg/blob/master/docs/examples/patching.md

**Comment by BillyONeal** [comment]
I'm closing this due to lack of response. Please feel free to create a new PR linking to this one which adds the appropriate patch if you would like to continue.

**Pull request closed without merging**